swarm/api/client/client.go

```diff
@@ -562,3 +563,89 @@ func (c *Client) MultipartUpload(hash string, uploader Uploader) (string, error)
 	}
 	return string(data), nil
 }
+
+// CreateResource creates a Mutable Resource with the given name and frequency, initializing it with the provided
+// data. Data is interpreted as multihash or not depending on the multihash parameter.
+// startTime=0 means "now"
+// Returns the resulting Mutable Resource manifest address that you can use to include in an ENS Resolver (setContent)
+// or reference future updates (Client.UpdateResource)
+func (c *Client) CreateResource(request *mru.Request) (string, error) {
+	responseStream, err := c.updateResource(request)
+	if err != nil {
+		return "", err
+	}
+	defer responseStream.Close()
+
+	body, err := ioutil.ReadAll(responseStream)
+	if err != nil {
+		return "", err
+	}
+
+	var manifestAddress string
+	if err = json.Unmarshal(body, &manifestAddress); err != nil {
+		return "", err
+	}
+	return manifestAddress, nil
+}
+
+// UpdateResource allows you to send a new version of your content
+func (c *Client) UpdateResource(request *mru.Request) error {
+	_, err := c.updateResource(request)
+	return err
+}
+
+func (c *Client) updateResource(request *mru.Request) (io.ReadCloser, error) {
+	body, err := request.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", c.Gateway+"/bzz-resource:/", bytes.NewBuffer(body))
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Body, nil
+
+}
+
+// GetResource returns a byte stream with the raw content of the resource
+// manifestAddressOrDomain is the address you obtained in CreateResource or an ENS domain whose Resolver
+// points to that address
+func (c *Client) GetResource(manifestAddressOrDomain string) (io.ReadCloser, error) {
+
+	res, err := http.Get(c.Gateway + "/bzz-resource:/" + manifestAddressOrDomain)
+	if err != nil {
+		return nil, err
+	}
+	return res.Body, nil
+
+}
+
+// GetResourceMetadata returns a structure that describes the Mutable Resource
+// manifestAddressOrDomain is the address you obtained in CreateResource or an ENS domain whose Resolver
+// points to that address
+func (c *Client) GetResourceMetadata(manifestAddressOrDomain string) (*mru.Request, error) {
+
+	responseStream, err := c.GetResource(manifestAddressOrDomain + "/meta")
```

Comment:
I don't agree with this. It is more complicated and obscure to use HTTP headers to specify the information you want to retrieve. Also more complicated for browser implementations and cors restrictions that need pre-flight when you specify weird headers.

Additionally, pasting in the browser address bar `http://localhost:8500/bzz-resource/12e1beb28d86ed828f9c38f064402e4fac9ca7b56dab9cf59103268a62a2b35f/meta` is much esier than having to open up a terminal and type `curl -H 'X-SWARM-METADATA' http://localhost:8500/bzz-resource/12e1beb28d86ed828f9c38f064402e4fac9ca7b56dab9cf59103268a62a2b35f` or something like that.

When designing an API, usually you specify the resource / identifier for the resource / subresource / identifier for subresource ...
In our case:
`/bzz-resource:/<hash>/meta` falls perfectly into this, since you want the "meta" subresource of the resource identified by `<hash>`

Reply:
So there are two things to this.

1) I doubt that it's mostly humans that will be performing requests for specific versions and periods. Thus I'm not sure if the argument of what's easier to put in curl or a browser address field is very relevant.

2) If we would want to allow resolving paths of underlying manifests through the mutable resource hash, then we would have to declare certain prefixes (period, version, meta) invalid. That's no good. I think a more natural comparison of expectations of how this should work would be _query strings_ rather than subpaths (which tend to be a `mod_rewrite` hack anyway). We already use query string for for example setting explicit content type on retrieval. So maybe that's what we should do here too?

Reply:
If we are going to be using paths to retrieve files, etc, I thing that'd be out of scope of `/bzz-resource:/` family of endpoints. I understand `/bzz-resource:/` as a way to manipulate the existence and lifetime of resources themselves.

Resolving paths and stuff like that shoud be `/bzz:/` or a new root level endpoint.

Reply:
What I mean is that I understand `bzz-resource:` as a low level API for resource lifecycle management, not something that supports higher level abstractions such as working with files and paths. That should be a higher layer service and should be on a different endpoint

Reply:
This is analogous to `bzz-raw:` providing low level access to things and `bzz:` providing the whole files and directories abstraction.

`bzz-raw:` does not care about files, just byte arrays. Doesn't even know what the name of the file was

Reply:
Yes, you are right, paths are irrelevant outside of `bzz:`. Still, this doesn't sit quite right by me. I think query string would be more natural. But to echo a common plea in this PR; let's make it a new PR.

Reply:
https://github.com/ethersphere/go-ethereum/issues/801

Reply:
I don't like these endpoints suffixing although both of you have fair points.
except @jpeletier the argument of having a `/meta` endpoint because headers are clumsy to type in a terminal is like buying run-flat tires just because you're too lazy to change a tire on your car. plus, a simple, one time written, end-to-end test will resolve clumsiness once and for all. as @nolash said this would be mostly programatically accessible and so url bars as considerations don't get any bonus points (by this logic we should also add a `GET bzz:/upload?data=` that accepts `data` as a base64 hex encoded data stream?).
@nolash i think we can land this as is, i'll have to clean it up later on anyway as part of the http refactoring.

Reply:
@justelad you bless it you bought it :money_with_wings:  Fine by me

Reply:
I think we are mixing discussions. The semantics of `bzz-resource:` and `bzz` are different. And those semantics affect whether it is appropriate or not to use HTTP headers depending on the operation.

Headers are to specify **how** you want the resource delivered, conditions on delivery and to carry the user's context (location, language, session). This affects delivery but does not determine the **what**.

To specify **what** you want, you use the URL with its path arrangements and query string possibilities.

Additionally, consider that use of custom headers will cause browsers to issue an extra preflight CORS check HTTP request, and that **sucks**.

Also note that some proxies may strip custom headers silently and mercilessly. Regarding proxies and intermediate caches, including the browser's: Responses to requests with the same URL but different headers could be cached as the same, so the header would need to be present in request and response, along with a `Vary` header in the response to indicate that the presence of a header affects the cache key. A mess.

Therefore: Avoid custom headers as much as possible. The easiness of using a straight URL was just a bonus point, not the main one.

Allow me to describe what I understand are the `bzz-resource:` and `bzz:` semantics, please let me know if I am not on the right page:

`bzz-resource:`: Manipulate resources at a low level. Create and update resources. Retrieve straight binary data. Refer to each resource with period+version in the URL, query string or part of the path.
`bzz:` static site / filesystem abstraction on top of Swarm.

For `bzz-resource:` I would:
* `POST /bzz-resource:/`: create / update resources
* `GET /bzz-resource:/{manifest}/meta`: obtain resource information
* `GET /bzz-resource:/{manifest}/period/{period}/version/{version}`: obtain raw octet stream of specific period, version. A querystring version of this would also be appropriate.
* `GET /bzz-resource:/{manifest}/period/{period}/`: obtain raw octect stream of latest version in period. A querystring version of this would also be appropriate.
* `GET /bzz-resource:/{manifest}/`: obtain raw content of latest version

Take into account that the `/meta` call will be needed by any client browser application prior to signing any MRU. Obtaining the resource's status information is indeed a `GET`. It doesn't make sense I get something completely different in the response just because I added a header saying I want the resource information.
Perhaps what is causing the confusion is the name `/meta`, maybe it should be renamed to `/info`.

With these semantics above, use of HTTP headers is not justified.

For `bzz:`:
* `GET /bzz:/{manifest}/path`: Resolve latest version of MRU as pointing to a manifest JSON... static site. That is, as it works today. Nothing new here.

Use HTTP headers to affect delivery of a particular resource. Thus, the same URL/site can be "viewed in the past" by adding, e.g.: `X-Swarm-Mru-Time: Wed, Apr 04 2018`. Swarm would then translate that to a period, version and retrieve that content, --or-- we could allow the user to specify `period`, `version` directly. We would also need to add the header to the response along with a `Vary: X-Swarm-Mru-Time` so that we don't confuse the browser cache or that of any intermediate proxy. 
Use of HTTP headers for this is OK, though you will need a browser extension to set that header for you if you'd like to browse a site in the past. The browser will be issuing preflights if headers are set on XMLHttpRequests.

Reply:
> Perhaps what is causing the confusion is the name /meta

For my part, no.

Anway, I think what you say about the **how** and **what** involving headers makes sense (although I don't think the cache and both sides headers really is such a mess). However, why not use _query strings_ is still a question.

Reply:
> However, why not use query strings is still a question

I am OK with query strings, this was more of a cosmetic taste. Also, be prepared to escape the insidious `&` in shell commands. ;-)

Reply:
> for bzz I prefer query strings when you go back in history.

My only concern here is that advanced static applications (Angular.JS, etc) may use query strings, so you could potentially affect the appication. This is a corner case, though.